[Lock] Split "StoreInterface" into multiple interfaces with less responsability

Simperfit opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | yes <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | TODO <!-- required for new features -->

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/roadmap):
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against branch 4.4.
 - Legacy code removals go to the master branch.
-->

We are removing the StoreInterface in order to split into multiple interface, it will help reduce de responsability of the StoreInterface.

Firstly, since not all stores needs to have all the methods of the StoreInterface, we can split this like this in order to limit the methods that are needed for each store.

Secondly, we add supportsX methods in order to avoid throwing exception when a store does not supports a feature it's easier an instance of the special interface or not, and it can return true/false on the support method.

Commit:
[Process] Allow writing portable "prepared" command lines

greg0ire committed:
[PhpUnitBridge] More accurate grouping

xabbuh committed:
accept floats for input="string" in NumberType

greg0ire committed:
[PhpunitBridge] Read environment variable from superglobals

The Dotenv component has recently been switched to using superglobals
instead of putenv(). Let us support both and give priority to
superglobals.

Closes #31857

TimoBakx committed:
[Messenger] Doctrine Connection find and findAll now correctly decode headers

alexsegura committed:
Add missing rendering of form help block.

derrabus committed:
[Mailer] Catch missing scheme in DSN.

Nyholm committed:
[Messenger] Remove DispatchAfterCurrentBusStamp when message is put on internal queue

Tobion committed:
[Messenger] make retry logic work without SentStamp

nicolas-grekas committed:
[HttpClient] fixing passing debug info to progress callback

Lctrs committed:
[Validator] Use LogicException for missing Property Access Component in comparison constraints

emr committed:
Turkish translation added to Form Component

Commit:
Fix link to documentation

alekitto committed:
[DebugBundle] fix register ReflectionCaster::unsetClosureFileInfo caster in var cloner service

alOneh committed:
[Mailgun Mailer] fixed issue when using html body

xabbuh committed:
fix typo

xabbuh committed:
sync `require-dev` and `conflict` constraints

fabpot committed:
bug #32136 [FrameworkBundle] sync `require-dev` and `conflict` constraints (xabbuh)

This PR was merged into the 4.3 branch.

Discussion
----------

[FrameworkBundle] sync `require-dev` and `conflict` constraints

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

The `conflict` section already prevents the Translation component 4.2 to be installed.

Commits
-------

47f9235568 sync `require-dev` and `conflict` constraints

fabpot committed:
bug #32094 [Validator] Use LogicException for missing Property Access Component in comparison constraints (Lctrs)

This PR was merged into the 4.2 branch.

Discussion
----------

[Validator] Use LogicException for missing Property Access Component in comparison constraints

| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | yes-ish
| New feature?  | no
| BC breaks?    | yes-ish
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

For consistency.

Commits
-------

6ac2316993 [Validator] Use LogicException for missing Property Access Component in comparison constraints

fabpot committed:
minor #32121 Fix link to documentation (AMartinNo1)

This PR was submitted for the master branch but it was merged into the 4.2 branch instead (closes #32121).

Discussion
----------

Fix link to documentation

| Q             | A
| ------------- | ---
| Branch?       | 4.4 for features / 3.4, 4.2 or 4.3 for bug fixes <!-- see below -->
| Bug fix?      | yes
| New feature?  |No, update link to VarExporter doc only
| BC breaks?    | no
| Deprecations? | no
| Test pass? | yes
| Fixed tickets | none
| License | MIT
| Doc PR | symfony/symfony-docs

Commits
-------

e0d2c58bc2 Fix link to documentation

fabpot committed:
bug #32125 [Form] accept floats for input="string" in NumberType (xabbuh)

This PR was merged into the 4.3 branch.

Discussion
----------

[Form] accept floats for input="string" in NumberType

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31905, #32124
| License       | MIT
| Doc PR        |

Commits
-------

2abf85599d accept floats for input="string" in NumberType

fabpot committed:
Merge branch '4.2' into 4.3

* 4.2:
  Fix link to documentation
  [Validator] Use LogicException for missing Property Access Component in comparison constraints

xabbuh committed:
fix accessing session bags

Tobion committed:
bug #32027 [Messenger] Remove DispatchAfterCurrentBusStamp when message is put on internal queue (Nyholm)

This PR was merged into the 4.3 branch.

Discussion
----------

[Messenger] Remove DispatchAfterCurrentBusStamp when message is put on internal queue

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #32009
| License       | MIT
| Doc PR        |

This will fix #32009.

Thank you @brpauwels for the report.

I consider it safe to remove the `DispatchAfterCurrentBusStamp` because its meaning disappear after we handled the "current bus".

T0: We add the stamp
T1: We put the envelope on an internal queue in `DispatchAfterCurrentBusMiddleware`
T2: We handle the current bus.
T3: We start processing our internal queue.

At T3 there we are "after current bus", that is why we dont need the stamp any more.

Commits
-------

91f1680b3f [Messenger] Remove DispatchAfterCurrentBusStamp when message is put on internal queue

xabbuh committed:
minor #32112 Turkish translation added to Form Component (emr)

This PR was submitted for the 4.4 branch but it was merged into the 3.4 branch instead (closes #32112).

Discussion
----------

Turkish translation added to Form Component

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

Commits
-------

a030e393c5 Turkish translation added to Form Component

xabbuh committed:
fix typos

xabbuh committed:
add test to avoid regressions

xabbuh committed:
fix mirroring directory into parent directory

nicolas-grekas committed:
bug #32129 [DebugBundle] fix register ReflectionCaster::unsetClosureFileInfo caster in var cloner service (alekitto)

This PR was merged into the 4.3 branch.

Discussion
----------

[DebugBundle] fix register ReflectionCaster::unsetClosureFileInfo caster in var cloner service

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

Non-existent class was checked by `method_exists` in Debug bundle.
This should fix (and correctly register) the caster while loading `DebugExtension` from `DebugBundle`

Commits
-------

860164ee7e [DebugBundle] fix register ReflectionCaster::unsetClosureFileInfo caster in var cloner service

nicolas-grekas committed:
minor #32140 [Filesystem] add test to avoid regressions (xabbuh)

This PR was merged into the 3.4 branch.

Discussion
----------

[Filesystem] add test to avoid regressions

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

Commits
-------

0d7d1f81bc add test to avoid regressions

nicolas-grekas committed:
bug #32083 [HttpClient] fixing passing debug info to progress callback (nicolas-grekas)

This PR was merged into the 4.3 branch.

Discussion
----------

[HttpClient] fixing passing debug info to progress callback

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

NativeHttpClient already has it.

Commits
-------

dc55cf826a [HttpClient] fixing passing debug info to progress callback

nicolas-grekas committed:
minor #31993 [Mailer] Catch missing scheme in DSN (derrabus)

This PR was merged into the 4.3 branch.

Discussion
----------

[Mailer] Catch missing scheme in DSN

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

The `Symfony\Component\Mailer\Transport::createTransport()` method parses and validates a passed DSN. I noticed that we never check if the DSN contains a valid scheme, but we always assume that the scheme is present in then parse result. If someone passes a DSN without a scheme to that method, they would almost certainly run into a PHP notice.

This PR makes sure that a scheme is present in the URL and throws a proper exception otherwise.

Commits
-------

3eba36c088 [Mailer] Catch missing scheme in DSN.

nicolas-grekas committed:
[HttpClient] fix dealing with 1xx informational responses

Tobion committed:
use proper return types in ErrorHandler and ArgumentResolver

Tobion committed:
Extract unrecoverable exception to interface

nicolas-grekas committed:
[HttpClient] fix timing measurements with NativeHttpClient

nikossvnk committed:
[Workflow] re-add workflow.definition tag to workflow services

Tobion committed:
[Messenger] fix retrying handlers using DoctrineTransactionMiddleware

Tobion committed:
bug #32053 [Messenger] No need for retry to require SentStamp (Tobion)

This PR was merged into the 4.3 branch.

Discussion
----------

[Messenger] No need for retry to require SentStamp

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |
| License       | MIT
| Doc PR        |

Fixes 2) in #32049
@weaverryan the SentStamp in the worker seems totally irrelevant. You always want to send messages back to the transport where they came from for retry. The only relevance I can potentially see is for the SyncTransport. Messages received async from worker might be routed to the SyncTransport. Using the SentStamp would redeliver the messages into the SyncTransport instead of the async. But that doesn't seem to have any use-case. I'm running the worker which handles the messages immediately. So basically there is no difference if they go to the Sync or Async transport from within the worker.

Commits
-------

0034dee641 [Messenger] make retry logic work without SentStamp

Simperfit committed:
[Lock] fix missing inherit docs in RedisStore

norkunas committed:
Fix type error

xabbuh committed:
collect called listeners information only once

bendavies committed:
revert #30525 due to performance penalty

nicolas-grekas committed:
[Security/Core] require libsodium >= 1.0.14

nicolas-grekas committed:
[Security/Core] Don't use ParagonIE_Sodium_Compat

fabpot committed:
bug #32156 [Workflow] re-add workflow.definition tag to workflow services (nikossvnk)

This PR was squashed before being merged into the 4.3 branch (closes #32156).

Discussion
----------

[Workflow] re-add workflow.definition tag to workflow services

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | #32155   <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        |

Add again the `workflow.definition` tag in workflow services.

Commits
-------

4df2dc5aaa [Workflow] re-add workflow.definition tag to workflow services

fabpot committed:
bug #32170 [Security/Core] Don't use ParagonIE_Sodium_Compat (nicolas-grekas)

This PR was merged into the 3.4 branch.

Discussion
----------

[Security/Core] Don't use ParagonIE_Sodium_Compat

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This is dead code:
https://github.com/paragonie/sodium_compat#features-excluded-from-this-polyfill

Commits
-------

61ea53d57f [Security/Core] Don't use ParagonIE_Sodium_Compat

fabpot committed:
bug #32169 [Security/Core] require libsodium >= 1.0.14 (nicolas-grekas)

This PR was merged into the 4.3 branch.

Discussion
----------

[Security/Core] require libsodium >= 1.0.14

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #32166
| License       | MIT
| Doc PR        | -

- bump libsodium to >=1.0.14
- Minimum opscost must be 3, as described in https://wiki.php.net/rfc/libsodium and in https://github.com/jedisct1/libsodium/releases/tag/1.0.15
- ParagonIE_Sodium_Compat [explicitly doesn't implement Argon2](https://github.com/paragonie/sodium_compat#features-excluded-from-this-polyfill), so it makes no sense to check for it.

Commits
-------

4fed5d3813 [Security/Core] require libsodium >= 1.0.14

fabpot committed:
bug #32154 [Messenger] fix retrying handlers using DoctrineTransactionMiddleware (Tobion)

This PR was merged into the 4.3 branch.

Discussion
----------

[Messenger] fix retrying handlers using DoctrineTransactionMiddleware

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |
| License       | MIT
| Doc PR        |

The retry logic only executes handlers that didn't fail using the HandledStamp. But in case of using the DoctrineTransactionMiddleware using several handlers, we need to remove the HandledStamp because those handlers got rolled back again.

Commits
-------

66c2e8483a [Messenger] fix retrying handlers using DoctrineTransactionMiddleware

dunglas committed:
[FrameworkBundle] Fix calling Client::getProfile() before sending a request

Simperfit committed:
[Lock] Split "StoreInterface" into multiple interfaces with less responsability

Tobion committed:
minor #32145 [Messenger] Extract unrecoverable exception to interface (Tobion)

This PR was merged into the 4.3 branch.

Discussion
----------

[Messenger] Extract unrecoverable exception to interface

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |
| License       | MIT
| Doc PR        |

Refactoring to extend messenger exceptions and extract unrecoverable exception into an interface. The exception is meant for people consuming messages. By using an interface, developers don't need to catch and rethrow or change their exception hierarchies. They can simply implement the interface in their existing exceptions to avoid unnecessary retries.

Commits
-------

6a2f4dc67a Extract unrecoverable exception to interface

Tobion committed:
[Form] remove comment about to-be-removed method as it is used in master by ButtonBuilder

fabpot committed:
bug #32163 [DoctrineBridge] Fix type error (norkunas)

This PR was merged into the 3.4 branch.

Discussion
----------

[DoctrineBridge] Fix type error

| Q             | A
| ------------- | ---
| Branch?       | 3.4 <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | -  <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | - <!-- required for new features -->

See discussion here: https://github.com/doctrine/DoctrineBundle/pull/984

> Having SET sql_mode=(SELECT REPLACE(@@sql_mode, 'ONLY_FULL_GROUP_BY', '')) query and when clicking on explain currently it throws:

> Argument 2 passed to Doctrine\DBAL\Connection::resolveParams() must be of the type array, null given, called in /vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php on line 911

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/roadmap):
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against branch 4.4.
 - Legacy code removals go to the master branch.
-->

Commits
-------

7b8ee3ece8 Fix type error

fabpot committed:
bug #32173 [FrameworkBundle] Fix calling Client::getProfile() before sending a request (dunglas)

This PR was merged into the 3.4 branch.

Discussion
----------

[FrameworkBundle] Fix calling Client::getProfile() before sending a request

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

Prevent throwing an error when calling getProfile before a request has been sent.

Commits
-------

9e6f4b2122 [FrameworkBundle] Fix calling Client::getProfile() before sending a request

fabpot committed:
bug #32164 [EventDispatcher] collect called listeners information only once (xabbuh)

This PR was merged into the 3.4 branch.

Discussion
----------

[EventDispatcher] collect called listeners information only once

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

Commits
-------

2ad32df6e0 collect called listeners information only once

fabpot committed:
Merge branch '3.4' into 4.2

* 3.4:
  [FrameworkBundle] Fix calling Client::getProfile() before sending a request
  Fix type error
  [Security/Core] Don't use ParagonIE_Sodium_Compat
  collect called listeners information only once
  add test to avoid regressions
  fix typos
  Turkish translation added to Form Component

fabpot committed:
Merge branch '4.2' into 4.3

* 4.2:
  [FrameworkBundle] Fix calling Client::getProfile() before sending a request
  Fix type error
  [Security/Core] Don't use ParagonIE_Sodium_Compat
  collect called listeners information only once
  add test to avoid regressions
  fix typos
  Turkish translation added to Form Component

fabpot committed:
minor #32177 [Form] remove comment about to-be-removed method  (Tobion)

This PR was merged into the 4.4 branch.

Discussion
----------

[Form] remove comment about to-be-removed method

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

it is used in master by ButtonBuilder, see https://github.com/symfony/symfony/pull/32160#issuecomment-505220078

Commits
-------

b80ce9ad18 [Form] remove comment about to-be-removed method as it is used in master by ButtonBuilder

fabpot committed:
bug #32165 revert #30525 due to performance penalty (bendavies)

This PR was merged into the 4.3 branch.

Discussion
----------

revert #30525 due to performance penalty

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |   <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        |

This reverts #30525, because it introduced a large performance penalty.

Here is the effect of that PR when I upgrade my Api Platform project from 4.2.9 to 4.3.1:
https://blackfire.io/profiles/compare/28bfbc61-3649-4896-bd03-7201239134cd/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=Symfony%5CComponent%5CVarExporter%5CInternal%5CExporter%3A%3Aexport%403&callname=main()

The reverted PR targeted master. This should go in 4.3?

Commits
-------

3d37cc98f6 revert #30525 due to performance penalty

jderusse commented on src/Symfony/Component/Lock/ExpirableStoreInterface.php:
```diff
@@ -0,0 +1,43 @@
+use Symfony\Component\Lock\Exception\NotSupportedException;
+
+/**
+ * StoreInterface defines an interface to manipulate a lock store.
```
`ExpirableStoreInterface`

jderusse commented on src/Symfony/Component/Lock/BlockingStoreInterface.php:
```diff
@@ -0,0 +1,38 @@
+use Symfony\Component\Lock\Exception\NotSupportedException;
+
+/**
+ * StoreInterface defines an interface to manipulate a lock store.
```
`BlockingStoreInterface`

fabpot committed:
minor #32161 [Lock] fix missing inherit docs in RedisStore (Simperfit)

This PR was merged into the 4.2 branch.

Discussion
----------

[Lock] fix missing inherit docs in RedisStore

| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | none   <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        |  <!-- required for new features -->

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/roadmap):
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against branch 4.4.
 - Legacy code removals go to the master branch.
-->

This is a missing inheritdoc in the RedisStore for the waitAndSave method.

Commits
-------

71eb8cfe99 [Lock] fix missing inherit docs in RedisStore

fabpot committed:
bug #32147 [HttpClient] fix timing measurements with NativeHttpClient (nicolas-grekas)

This PR was merged into the 4.3 branch.

Discussion
----------

[HttpClient] fix timing measurements with NativeHttpClient

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Align to what curl does.

Commits
-------

c5c3332400 [HttpClient] fix timing measurements with NativeHttpClient

jderusse commented on src/Symfony/Component/Lock/Lock.php:
```diff
@@ -36,13 +36,17 @@ final class Lock implements LockInterface, LoggerAwareInterface
+        if ($store instanceof StoreInterface) {
+            @trigger_error(sprintf('"%s" is deprecated since Symfony 4.4 and has been splitted into "%s", "%s", "%s".', StoreInterface::class, PersistStoreInterface::class, ExpirableStoreInterface::class, BlockingStoreInterface::class), E_USER_DEPRECATED);
+        }
+
```
You should assert here that the `$store` is an `PersistStoreInterface`.

jderusse commented on src/Symfony/Component/Lock/Lock.php:
```diff
@@ -36,13 +36,17 @@ final class Lock implements LockInterface, LoggerAwareInterface
-     * @param float|null     $ttl         Maximum expected lock duration in seconds
-     * @param bool           $autoRelease Whether to automatically release the lock or not when the lock instance is destroyed
+     * @param Key                                                                  $key         Resource to lock
+     * @param PersistStoreInterface|BlockingStoreInterface|ExpirableStoreInterface $store       Store used to handle lock persistence
```
IMHO it's not a "or" `|`, it should be a `PersistStoreInterface` + optional interfaces

jderusse commented on src/Symfony/Component/Lock/Lock.php:
```diff
@@ -69,7 +73,7 @@ public function __destruct()
     {
         try {
-            if ($blocking) {
+            if ($blocking && $this->store instanceof BlockingStoreInterface) {
```
`&& $this->store->supportsWaitAndSave()`

jderusse commented on src/Symfony/Component/Lock/Lock.php:
```diff
@@ -121,7 +125,9 @@ public function refresh($ttl = null)
         try {
             $this->key->resetLifetime();
-            $this->store->putOffExpiration($this->key, $ttl);
+            if ($this->store instanceof ExpirableStoreInterface) {
```
`&& $this->store->supportsPutOffExpiration`

fabpot committed:
bug #32137 [HttpFoundation] fix accessing session bags (xabbuh)

This PR was merged into the 3.4 branch.

Discussion
----------

[HttpFoundation] fix accessing session bags

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30682
| License       | MIT
| Doc PR        |

Commits
-------

7a4570dcac fix accessing session bags

jderusse commented on src/Symfony/Component/Lock/Store/CombinedStore.php:
```diff
@@ -181,4 +184,20 @@ public function exists(Key $key)
+     */
+    public function supportsWaitAndSave(): bool
+    {
+        return true;
```
`false`

jderusse commented on src/Symfony/Component/Lock/Store/CombinedStore.php:
```diff
@@ -181,4 +184,20 @@ public function exists(Key $key)
+     */
+    public function supportsPutOffExpiration(): bool
+    {
+        return true;
```
check if at least 1 store supports expiration.

At line  118 you should also check if the stores supports expiration and throws an exception if none of theme support it

jderusse commented on src/Symfony/Component/Lock/Store/MemcachedStore.php:
```diff
@@ -21,7 +22,7 @@
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-class MemcachedStore implements StoreInterface
+class MemcachedStore implements ExpirableStoreInterface
```
`PersistStoreInterface`

fabpot committed:
minor #32143 use proper return types in ErrorHandler and ArgumentResolver (Tobion)

This PR was merged into the 4.4 branch.

Discussion
----------

use proper return types in ErrorHandler and ArgumentResolver

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | tiny
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |
| License       | MIT
| Doc PR        |

Found those things while reviewing #31996 which missed some return types due to using `return` instead of `return null`.
It's part of fixing #17201 (due to #10717). See also #30869 that somebody was stumbling over.

Commits
-------

2f9121b74d use proper return types in ErrorHandler and ArgumentResolver

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -99,4 +104,20 @@ public function exists(Key $key)
+     */
+    public function supportsPutOffExpiration(): bool
+    {
+        return true;
```
`return $this->decorated instanceof ExpirableStoreInterface && $this->decorated->supportsPutOffExpiration();`

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -99,4 +104,20 @@ public function exists(Key $key)
+     */
+    public function supportsWaitAndSave(): bool
+    {
+        return true;
```
`$this->decorated instanceof BlockingStoreInterface && $this->decorated->supportsWaitAndSave()`

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -81,7 +84,9 @@ public function waitAndSave(Key $key)
     public function putOffExpiration(Key $key, $ttl)
     {
-        $this->decorated->putOffExpiration($key, $ttl);
+        if ($this->decorated instanceof ExpirableStoreInterface && $this->decorated->supportsPutOffExpiration()) {
```
:+1:

Should do the same in `waitAndSave`

jderusse commented on src/Symfony/Component/Lock/ExpirableStoreInterface.php:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @author Hamza Amrouche <hamza.simperfit@gmail.com>
+ */
+interface ExpirableStoreInterface
```
`Expiring`?

fabpot committed:
bug #32138 [Filesystem] fix mirroring directory into parent directory (xabbuh)

This PR was merged into the 4.3 branch.

Discussion
----------

[Filesystem] fix mirroring directory into parent directory

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #32089
| License       | MIT
| Doc PR        |

Commits
-------

b58a806340 fix mirroring directory into parent directory

jderusse commented on src/Symfony/Component/Lock/Store/SemaphoreStore.php:
```diff
@@ -98,10 +101,18 @@ public function delete(Key $key)
 
     /**
-     * {@inheritdoc}
+     * Extends the ttl of a resource.
```
Is this docblock needed? What's about other stores?

Simperfit replied:
I need to add it on other stores

jderusse commented on src/Symfony/Component/Lock/Store/CombinedStore.php:
```diff
@@ -26,7 +28,7 @@
  *
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-class CombinedStore implements StoreInterface, LoggerAwareInterface
```
This is a BC break for code like:
```
class Service 
{
  public function __construct(StoreInterface $store)
  {
  }
}
```

Could be fixed in FrameworkExtension by decorating the store with a `LegacyStore` which trigger the deprecation

fabpot committed:
bug #32141 [HttpClient] fix dealing with 1xx informational responses (nicolas-grekas)

This PR was merged into the 4.3 branch.

Discussion
----------

[HttpClient] fix dealing with 1xx informational responses

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

I never had a look at 1xx status codes until today.
This PR fixes reading them when using curl.

If one wonders:
- `NativeHttpClient` uses `fopen()`, which skips informational parts as allowed by the HTTP spec and doesn't give any way to access their response headers.
- `CurlHttpClient` allows reading informational responses using the progress callback or via the getInfo() method. That's the way if you need to implement e.g. HTTP 103 early hints.

Commits
-------

412411d795 [HttpClient] fix dealing with 1xx informational responses

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -37,7 +40,7 @@ class RetryTillSaveStore implements StoreInterface, LoggerAwareInterface
      * @param int            $retryCount Maximum amount of retry
      */
-    public function __construct(StoreInterface $decorated, int $retrySleep = 100, int $retryCount = PHP_INT_MAX)
+    public function __construct(PersistStoreInterface $decorated, int $retrySleep = 100, int $retryCount = PHP_INT_MAX)
```
BC Break. you should allow both of theme

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -81,7 +84,9 @@ public function waitAndSave(Key $key)
-        $this->decorated->putOffExpiration($key, $ttl);
+        if ($this->decorated instanceof ExpirableStoreInterface && $this->decorated->supportsPutOffExpiration()) {
+            $this->decorated->putOffExpiration($key, $ttl);
+        }
```
else throw an exception

jderusse reviewed (commented):
First review :-)

fabpot committed:
feature #32126 [Process] Allow writing portable "prepared" command lines (Simperfit)

This PR was merged into the 4.4 branch.

Discussion
----------

[Process] Allow writing portable "prepared" command lines

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | #23778  <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/11802 <!-- required for new features -->

Hey here, it's me, again !

I've talked with @nicolas-grekas and he gave me a new way of writing this feature that will not be a problem for people using things like {{ toto }} since we are using the linux style of using envvar, the replaced arguments is only replaced when using windows.

This should not break anything and work as expected!

see https://github.com/symfony/symfony/pull/24763

This makes `"$FOO"` work seamlessly on Linux and on Windows to reference an env var (that can be provided when calling e.g. the "run" method)

Commits
-------

3f8354f58f [Process] Allow writing portable "prepared" command lines

fabpot committed:
bug #31972 Add missing rendering of form help block. (alexsegura)

This PR was merged into the 4.2 branch.

Discussion
----------

Add missing rendering of form help block.

| Q             | A
| ------------- | ---
| Branch?       | 4.2 <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| License       | MIT

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/roadmap):
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against branch 4.4.
 - Legacy code removals go to the master branch.
-->

When using `bootstrap_3_layout.html.twig` form theme, the help blocks are not rendered.

```twig
{% form_theme form 'bootstrap_3_layout.html.twig' %}

{{ form_start(form) }}
  {{ form_row(form.some_field) }} {# <-- no help text is rendered #}
{{ form_end(form) }}
```

Commits
-------

ff5517e554 Add missing rendering of form help block.

fabpot committed:
bug #31966 [Messenger] Doctrine Connection find and findAll now correctly decode headers (TimoBakx)

This PR was merged into the 4.3 branch.

Discussion
----------

[Messenger] Doctrine Connection find and findAll now correctly decode headers

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31916
| License       | MIT
| Doc PR        | N/A

The Doctrine transport `Connection` class methods `find` and `findAll` did not JSON-decode the headers of the envelope after retrieving it from Doctrine. The `get` method, however, did this correctly.

I added two tests to verify the results of `find` and `findAll` and then added the JSON-decoding to the `Connection` class. I moved the JSON-decoding to a separate method to avoid duplicate code.

Commits
-------

3aec2acce5 [Messenger] Doctrine Connection find and findAll now correctly decode headers

fabpot committed:
fixed CS

fabpot committed:
bug #31730 [PhpUnitBridge] More accurate grouping (greg0ire)

This PR was squashed before being merged into the 4.3 branch (closes #31730).

Discussion
----------

[PhpUnitBridge] More accurate grouping

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

Sometimes, you cannot easily know if code was written by a vendor or
directly in the application, for instance if the code comes from a file
in the cache. In that case, it is better not to classify the deprecation
as direct or indirect.

@jmsche please test this on your application when you can, I think you might be having that issue.

Commits
-------

d9f0ba386a [PhpUnitBridge] More accurate grouping

nicolas-grekas committed:
[Bridge/PhpUnit] Fix PHP5.5 compat

fabpot committed:
bug #32131 [Mailgun Mailer] fixed issue when using html body (alOneh)

This PR was merged into the 4.3 branch.

Discussion
----------

[Mailgun Mailer] fixed issue when using html body

| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

I tested the `symfony/mailgun-mailer` and get an issue when using the `api` scheme with a templated email cause we try to manipulate a stream whereas the `Symfony\Component\Mime\Email::getHtmlBody()` could return also a string (in my case it is one).

The issue :

```
stream_get_meta_data() expects parameter 1 to be resource, string given
```

Commits
-------

afbefe131b [Mailgun Mailer] fixed issue when using html body

nicolas-grekas committed:
bug #31954 [PhpunitBridge] Read environment variable from superglobals (greg0ire)

This PR was merged into the 4.3 branch.

Discussion
----------

[PhpunitBridge] Read environment variable from superglobals

| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31857
| License       | MIT
| Doc PR        | n/a

The Dotenv component has recently been switched to using superglobals
instead of putenv(). Let us support both and give priority to
superglobals.

Commits
-------

88cfcb536d [PhpunitBridge] Read environment variable from superglobals

nicolas-grekas committed:
Merge branch '3.4' into 4.2

* 3.4:
  fix accessing session bags

nicolas-grekas committed:
Merge branch '4.2' into 4.3

* 4.2:
  [Lock] fix missing inherit docs in RedisStore
  fix accessing session bags
  Add missing rendering of form help block.

nicolas-grekas committed:
Merge branch '4.3' into 4.4

* 4.3: (34 commits)
  [PhpunitBridge] Read environment variable from superglobals
  [Bridge/PhpUnit] Fix PHP5.5 compat
  [PhpUnitBridge] More accurate grouping
  fixed CS
  Extract unrecoverable exception to interface
  [FrameworkBundle] Fix calling Client::getProfile() before sending a request
  Fix type error
  [Security/Core] require libsodium >= 1.0.14
  [Workflow] re-add workflow.definition tag to workflow services
  [Security/Core] Don't use ParagonIE_Sodium_Compat
  revert #30525 due to performance penalty
  collect called listeners information only once
  [Lock] fix missing inherit docs in RedisStore
  [Messenger] fix retrying handlers using DoctrineTransactionMiddleware
  [Mailgun Mailer] fixed issue when using html body
  [HttpClient] fix timing measurements with NativeHttpClient
  [HttpClient] fix dealing with 1xx informational responses
  add test to avoid regressions
  fix mirroring directory into parent directory
  fix typos
  ...

nicolas-grekas committed:
Fix typo

nicolas-grekas committed:
Bump phpunit-bridge

nicolas-grekas committed:
Merge branch '3.4' into 4.2

* 3.4:
  Bump phpunit-bridge

nicolas-grekas committed:
Merge branch '4.2' into 4.3

* 4.2:
  Bump phpunit-bridge

nicolas-grekas committed:
Merge branch '4.3' into 4.4

* 4.3:
  Bump phpunit-bridge

nicolas-grekas committed:
[PhpUnitBridge] Fix tests

nicolas-grekas committed:
Merge branch '4.3' into 4.4

* 4.3:
  [PhpUnitBridge] Fix tests

nicolas-grekas committed:
Revert "bug #31730 [PhpUnitBridge] More accurate grouping (greg0ire)"

This reverts commit 64b68d49226b7095aec69f3a2b86b17f3d049e79, reversing
changes made to 3cd795fddeb3466f7be195ffd6108577aa639fdf.

nicolas-grekas committed:
Merge branch '4.3' into 4.4

* 4.3:
  Revert "bug #31730 [PhpUnitBridge] More accurate grouping (greg0ire)"

nicolas-grekas committed:
Revert "Revert "bug #31730 [PhpUnitBridge] More accurate grouping (greg0ire)""

This reverts commit 4814fd3a4e3c7a668e063e1d4d2cf10e8d1b0d23.

nicolas-grekas committed:
Reject phpunit-bridge v5 for now

nicolas-grekas committed:
Merge branch '4.3' into 4.4

* 4.3:
  Reject phpunit-bridge v5 for now
  Revert "Revert "bug #31730 [PhpUnitBridge] More accurate grouping (greg0ire)""

jderusse commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1600,9 +1607,14 @@ private function registerLockConfiguration(array $config, ContainerBuilder $cont
+                $legacyDefinition = new ChildDefinition('lock.store.legacy.abstract');
+                $legacyDefinition->replaceArgument(0, $combinedDefinition);
+
+                $container->setDefinition('lock.'.$resourceName.'.store', $legacyDefinition);
```
`$container->setDefinition('lock.'.$resourceName.'.persist_store', $definition);`

jderusse commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1625,9 +1637,13 @@ private function registerLockConfiguration(array $config, ContainerBuilder $cont
                 $container->setAlias('lock.factory', new Alias('lock.'.$resourceName.'.factory', false));
                 $container->setAlias('lock', new Alias('lock.'.$resourceName, false));
                 $container->setAlias(StoreInterface::class, new Alias('lock.store', false));
+                $container->setAlias(PersistStoreInterface::class, new Alias('lock.persist_store', false));
```
`lock.persist_store` is not defined

should deprecate `lock.store` and `lock.store.*`

jderusse commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/lock.xml:
```diff
@@ -24,6 +24,12 @@
             <argument type="service" id="lock.strategy.majority" /> <!-- Strategy -->
         </service>
 
+        <service id="lock.store.legacy.abstract" class="Symfony\Component\Lock\Store\LegacyStore" abstract="true">
```
not used?

jderusse commented on src/Symfony/Component/Lock/Factory.php:
```diff
@@ -26,8 +26,11 @@ class Factory implements LoggerAwareInterface
+    public function __construct($store)
     {
+        if ($store instanceof StoreInterface) {
+            @trigger_error(sprintf('"%s" is deprecated since Symfony 4.4 and has been splitted into "%s", "%s", "%s".', StoreInterface::class, PersistStoreInterface::class, ExpiringStoreInterface::class, BlockingStoreInterface::class), E_USER_DEPRECATED);
```
This error should have been laready triggered when user autoload the interface

jderusse commented on src/Symfony/Component/Lock/Store/CombinedStore.php:
```diff
@@ -104,9 +118,10 @@ public function putOffExpiration(Key $key, $ttl)
         $storesCount = \count($this->stores);
         $expireAt = microtime(true) + $ttl;
-
+        $atLeastOneStoreSupportsExpiration = $this->supportsPutOffExpiration();
```
I would have trigger exceptiojn her.
```
 if (false === $atLeastOneStoreSupportsExpiration) {
    throw new NotSupportedException();
 }
```

jderusse commented on src/Symfony/Component/Lock/Store/LegacyStore.php:
```diff
@@ -0,0 +1,122 @@
+use Symfony\Component\Lock\Strategy\StrategyInterface;
+
+/**
+ * CombinedStore is a StoreInterface implementation able to manage and synchronize several StoreInterfaces.
```
nop

jderusse commented on src/Symfony/Component/Lock/Store/LegacyStore.php:
```diff
@@ -0,0 +1,122 @@
+/**
+ * CombinedStore is a StoreInterface implementation able to manage and synchronize several StoreInterfaces.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
```
nop

jderusse commented on src/Symfony/Component/Lock/Store/LegacyStore.php:
```diff
@@ -0,0 +1,122 @@
+     */
+    public function supportsPutOffExpiration(): bool
+    {
+        if ($this->decorated instanceof ExpiringStoreInterface && $this->decorated->supportsPutOffExpiration()) {
```
`return $this->decorated instanceof ExpiringStoreInterface && $this->decorated->supportsPutOffExpiration();`

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -37,8 +42,16 @@ class RetryTillSaveStore implements StoreInterface, LoggerAwareInterface
+            @trigger_error(sprintf('"%s" is deprecated since Symfony 4.4 and has been splitted into "%s", "%s", "%s".', StoreInterface::class, PersistStoreInterface::class, ExpiringStoreInterface::class, BlockingStoreInterface::class), E_USER_DEPRECATED);
+        }
+
+        if ($decorated instanceof PersistStoreInterface || $decorated instanceof StoreInterface) {
```
`if NOT`

jderusse commented on src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:
```diff
@@ -59,29 +72,37 @@ public function save(Key $key)
-        do {
-            try {
-                $this->decorated->save($key);
+        if ($this->decorated instanceof BlockingStoreInterface && $this->decorated->supportsWaitAndSave()) {
```
use early return
```
if (!$this->decorated instanceof BlockingStoreInterface || !$this->decorated->supportsWaitAndSave()) {
  throw new NotSupportedException();
}
```
rest of code should net change

jderusse reviewed (commented):
another pass
